src/perf/isolates.md

```diff
@@ -0,0 +1,355 @@
+even if mutated on that isolate,
+doesn't change the value on the original isolate.
+
+The only objects that aren't copied when passed to an isolate
+are immutable objects that can't be changed anyway,
+such a String or an unmodifiable byte.
+When you pass an immutable object between isolates,
+it is transferred, rather than copied,
+for better performance.
+Because immutable objects can't be updated,
+this effectively retains the actor model behavior.
+
+An exception to this rule is
+when an isolate exits when it sends a message using the `Isolate.exit` method.
+Because the sending isolate won't exist after sending the message,
+it can pass the reference to the message while
+ensuring that only one isolate can access the message.
+
+The two lowest-level primitives that send messages are `SendPort.send`,
+which makes a copy of a mutable message as it sends,
+and `Isolate.exit`,
+which sends the reference to the message.
+Both `Isolate.run` and `compute`
+use `Isolate.exit` under the hood.
+
+## Short-lived isolates
+
+The easiest way to move a process to an isolate in Flutter is with
+the `Isolate.run` method.
+This method spawns an isolate,
+passes a callback to the spawned isolate to start some computation,
+returns a value from the computation,
+and then shuts the isolate down when the computation is complete.
+This all happens concurrently with the main isolate,
+and doesn't block it.
+
+![Isolate diagram]({{site.url}}/assets/images/docs/development/concurrency/isolate-bg-worker.png){:width="50%"}
+
+The `Isolate.run` method requires a single argument,
+a callback function,
+that is run on the new isolate.
+This callback's function signature must have exactly
+one required, unnamed argument.
+When the computation completes,
+it returns the callback's value back to the main isolate,
+and exits the spawned isolate.
+
+For example,
+consider this code that loads a large JSON blob from a file,
+and converts that JSON into custom Dart objects.
+If the json decoding process wasn't off loaded to a new isolate,
+this method would cause the UI to
+become unresponsive for several seconds.
+
+<?code-excerpt "lib/main.dart (isolate-run)"?>
+```dart
+// Produces a list of 211,640 photo objects.
+// (The JSON file is ~20MB.)
+Future<List<Photo>> getPhotos() async {
+  String jsonString = await rootBundle.loadString('assets/photos.json');
+
+  final dynamic photos = await Isolate.run(() {
+    final photoData = jsonDecode(jsonString);
+    return photoData.map((dynamic element) {
+      final data = element as Map<String, dynamic>;
+      return Photo.fromJson(data);
+    }).toList();
+  });
+
+  return photos as List<Photo>;
+}
+```
+For a complete walkthrough of using Isolates to 
+parse JSON in the background, see [this cookbook recipe][].
+
+[this cookbook recipe]: {{site.url}}/cookbook/networking/background-parsing
+
+### Isolates on the web
+
+Dart doesn't support isolates on web platforms.
+If you're targeting the web with your Flutter app,
```

Comment:
Having to recommend `compute()` on the web, where it runs things on the main thread anyway, seems kind of defeating the point of offloading work onto a different thread/isolate.

Instead, you can refer to the `WebWorker` API, which is actually quite similar to the Isolate API (start/stop a WebWorker, passing messages & transferable by-reference data types).

We have first-party support in [package:web](https://github.com/dart-lang/web/blob/e99d53716709d1cbce5533a7f6d036a8a5e19f68/lib/src/dom/html.dart#L4782) You could also link to https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API

Perhaps a code sample where you write the photo sample with a WebWorker could be nice to have?

Reply:
Overwhelmingly, the most requested addition to our docs about isolates is being clear about the limitations. We have resources sprinkled across the web that use `compute` in example code, and it's important that we're explicit about what that means. 

I think documenting WebWorkers is beyond the scope of this update.

Reply:
I agree we don't need to document web workers here, but we do need to be careful about how we write about `compute` and its behavior on web.

Some people incorrectly see it as "using `compute` instead of `Isolate.run` allows the same behavior to work on web".

I wonder if we should just always talk about and default to `Isolate.run`, but then in a small note or similar put forward `compute` as a way that allows the same code to **compile** for web without changes, but highlight that it does **not work** the same. That way it's a more explicit choice they have to make after they know the differences.